Use Catch2v3 for the unit test suite

alt-graph opened the pull request:
This moves from the old header-only Catch2 version 2.x included in GUL14 to the newer 3.x family of the unit test framework. Accordingly, the header files that need to be included change.

alt-graph committed:
Use Catch2v3 for the unit test suite

This moves from the old header-only Catch2 version 2.x included in GUL14
to the newer 3.x family of the unit test framework. Accordingly, the
header files that need to be included change.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Meson: Add wrap file for Catch2

[why]
This way we can avoid having to install it for CI runs.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
CI: Add Ubuntu 24, update upload-artifact action to v4

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
CI: Use Meson 1.7.0

[why]
We need modern support for wrap files without explicitly requesting a
fallback in the meson.build. So why not go to an up-to-date version?

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
This currently builds all of Catch2v3 in the CI runs. Once we drop Ubuntu 20.04, we can just pre-install the `catch2` package.

Finii commented on .github/workflows/build_and_run_tests.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
     strategy:
       matrix:
-        os: [ ubuntu-20.04, ubuntu-22.04 ]
+        os: [ ubuntu-20.04, ubuntu-22.04, ubuntu-24.04 ]
```
I would drop 22.04, like we just did with `taskolib`.

alt-graph replied:
I'll leave it in for now. It adds one more compiler version to the mix and doesn't cost anything. I would rather get rid of Ubuntu 20 as fast as possible.

Finii commented on .github/workflows/build_and_run_tests.yml:
```diff
@@ -45,7 +45,7 @@ jobs:
       with:
         repository: "mesonbuild/meson"
-        tag: "0.53.2"
+        tag: "1.7.0"
```
I'm not sure why we do this?
 Usually we use the lowest version we promise to support (promise in `meson.build`), and that did not change.

On the contrary we say `meson_version : '>=0.49'`.

Whatever, I usually use `HEAD` of Meson locally, and if we use the lowest allowed version in the CI we get a nice test range.

Additionally I think we should use the Meson version that is system-package-available on the test-OS (if it suffices).

In this way we would use 0.53.2 on 20.04 and 1.3.2 on 24.04.

alt-graph replied:
The reason is that `catch2`, when built as a fallback via the wrap file, requires meson >0.54. Therefore, Ubuntu 20 is a problem.

Finii replied:
Would it make sense to use 0.54.1 here?

alt-graph replied:
If we do that, we also have to add the explicit fallback arguments in tests/meson.build. More modern Meson versions can read the information directly from the wrap file. I am inclined to stay with a modern version here...

Finii commented on tests/meson.build:
```diff
@@ -11,7 +11,10 @@ test_src = files(
         test_src,
-        dependencies : libgit4cpp_dep,
+        dependencies : [
+            dependency('catch2'),
```
`catch2` could also be catch2v2. We should add a version number.

Probably `>>3.4.1` ?

alt-graph replied:
I have added `version: '>3.4.0'`. This is in line with the declaration in the `control` file.

Finii reviewed (changes requested):
`debian/control` has no Build-Depends on Catch2v3.

alt-graph committed:
Debian: Add build-depends on catch2 >3.4.0

[why]
We now require Catch2 v3 from an external package for building the unit
test suite.

Proposed-by: Fini Jastrow <fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
> `debian/control` has no Build-Depends on Catch2v3.

Good point, added.

alt-graph committed:
Meson: Require catch2 >3.4.0 for building tests

[why]
The precompiled library form of the Catch2 framework that we use was
introduced in Catch2 3.4.0-1 or something like that.

Proposed-by: Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

Finii commented on debian/control:
```diff
@@ -3,7 +3,7 @@ Section: devel
 Maintainer: Jannik Woehnert <jannik.woehnert@desy.de>
 Standards-Version: 4.6.0
-Build-Depends: dev-doocs-libgul14 | libgul14-dev, meson (>=0.50.0), pkg-config, debhelper-compat (=12), libgit2-dev
+Build-Depends: dev-doocs-libgul14 | libgul14-dev, meson (>=0.50.0), pkg-config, debhelper-compat (=12), catch2 (>>3.4.0), libgit2-dev
```
> `Add build-depends on catch2 >3.4.0`

Hmm, we have a problem here, out there is just this one package that is an in-between one:

`3.4.0-1build1`: Only on Ubuntu, and there on 24.04 and 24.10.

1. The header only catch2 packages were called `catch2`
2. The library v3 version is _destined_ to be called `libcatch2`
3. There is a transitional package transmogrifying `catch2` to `libcatch2`

On Ubuntu systems after 24.10 it is possible to install `libcatch2` and not have `catch2` installed.

On Debian we do not have this break, and Catch2v3 is always available as `libcatch2`. I guess that Ubuntu missed (or did not anticipate) that naming in Debian, and later on needed to adapt :grimacing: 

So probably what we really need, sadly, is :sob: 

        catch2 (>=3.4.0) | libcatch2

The pkgconfig name is `catch2` for both.

I have no clue what the rationale was at Debian, probably because there are people out there still needing the header only one, and to not-break that they invented a new name, which is quite a good idea.
But Ubuntu blundered here :disappointed: 

:thinking: RedHat's name for the package is `catch` :unamused: 

![image](https://github.com/user-attachments/assets/4f1cad3d-f516-4e8b-afad-a2c0cc20ef66)

![image](https://github.com/user-attachments/assets/3910e277-affe-44a5-81b7-be4785f6c465)

alt-graph replied:
Changed the control file accordingly, see [670e263](https://github.com/taskolib/libgit4cpp/pull/26/commits/670e2632bbc8c201cf9aac3a995846dd89950b42).

Finii commented:
Sorry this seems all so complicated, but I believe it is good to get this right at one place before we reuse our 'pattern' throughout doocs.

alt-graph commented:
> Would it make sense to use 0.54.1 here?

If we do that, we also have to add the explicit fallback arguments in `tests/meson.build`. More modern Meson versions can read the information directly from the wrap file. I am inclined to stay with a modern version here...

alt-graph committed:
Debian: Change catch2 build dependency

This changes the build-depends clause from  "catch2 (>>3.4.0)" to
"catch2 (>=3.4.0) | libcatch2". Rationale (Fini):

"
Hmm, we have a problem here, out there is just this one package that
is an in-between one:

3.4.0-1build1: Only on Ubuntu, and there on 24.04 and 24.10.

    The header only catch2 packages were called catch2
    The library v3 version is destined to be called libcatch2
    There is a transitional package transmogrifying catch2 to libcatch2

On Ubuntu systems after 24.10 it is possible to install libcatch2 and
not have catch2 installed.

On Debian we do not have this break, and Catch2v3 is always available
as libcatch2.
"

Co-authored-by: Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

Finii commented on tests/meson.build:
```diff
@@ -11,7 +11,10 @@ test_src = files(
         test_src,
-        dependencies : libgit4cpp_dep,
+        dependencies : [
+            dependency('catch2', version : '>3.4.0'),
```
This is now not-in-sync with the `debian/control` anymore which requires `>= 3.4.0`

alt-graph replied:
Changed to `>=`. Thanks a lot, @Finii!

Finii reviewed (approved):
I guess this is good to go.
The one comment has no practical impact I believe, because we have 

... we have version `3.4.0-blahfasel` which is of course `>` than `3.4.0` ;-)

alt-graph committed:
Meson: Require catch2 >=3.4.0 (instead of >3.4.0)

[why]
To bring the requirement in sync with the one listed in the Debian
control file.

Proposed-by: Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>
